src/arch/MovieTexture/MovieTexture_FFMpeg.h

```diff
@@ -48,8 +70,26 @@ class MovieDecoder_FFMpeg: public MovieDecoder
 	void Close();
 	void Rewind();
 
-	void GetFrame( RageSurface *pOut );
-	int DecodeFrame( float fTargetTime );
+	// This draws a frame from the buffer onto the provided RageSurface.
+	// Returns true if returning the last frame in the movie.
+	bool GetFrame(RageSurface* pOut);
+
+	// Decode a single frame.  Return -2 on cancel, -1 on error, 0 on EOF, 1 if we have a frame.
+	int DecodeNextFrame();
+
+	// Decode the entire movie.
+	// If we let this decode as fast as possible, it could come at the expense
+	// of gameplay performance. Since dropping frames is undesirable, an
+	// artificial rate limit is introduced.
+	//
+	// Given that most movies will display at 30fps or 60fps, decoding at a
+	// speed of 1000fps should be more than sufficient to ensure we never
```

Comment:
It's not clear from the nearby code where this value of 1000fps is coming from? (at least to me)

Reply:
There's a few nuances worth discussing here,

On the subject of 1000fps, if your question is "where did that number come from," I pulled it out of thin air. Empirically, my machine seemed fine decoding at that speed, though a lower powered machine may struggle--that needs a bit more fine tuning (or a setting perhaps for decoding speed in the options). Realistically, we could pick any rate that's faster than display without much worry.

If your question is, "where is this rate applied" then that's in `MovieTexture_FFMpeg.cpp::DecodeMovie`, right around here:

https://github.com/itgmania/itgmania/pull/361/commits/b100578288230f5707df96ea38c86d6a34b550db#diff-0aec22a6fe4107b7925c7d7fe6ab2e7f043611b830dd5a397183216e77b0dac7R257

The code here is basically saying, "if we finished decoding before 1ms, then sleep till then." So 1ms per frame == 1000fps decoding. Now I'd like to switch this to a RageTimer too (right now it's using chrono), but I'm relying on `std::this_thread::sleep_until`, which takes a chrono time input. Perhaps I should be using a `RageThread` instead (or convert from RageTimer->chrono time), I can do a bit more research there.

Reply:
Just to be sure, the FPS in this case relate to FPS of the video file to be decoded and played, not FPS rendered by the game itself, right?

Reply:
> There's a few nuances worth discussing here,
> 
> On the subject of 1000fps, if your question is "where did that number come from," I pulled it out of thin air. Empirically, my machine seemed fine decoding at that speed, though a lower powered machine may struggle--that needs a bit more fine tuning (or a setting perhaps for decoding speed in the options). Realistically, we could pick any rate that's faster than display without much worry.
> 
> If your question is, "where is this rate applied" then that's in `MovieTexture_FFMpeg.cpp::DecodeMovie`, right around here:
> 
> [b100578#diff-0aec22a6fe4107b7925c7d7fe6ab2e7f043611b830dd5a397183216e77b0dac7R257](https://github.com/itgmania/itgmania/commit/b100578288230f5707df96ea38c86d6a34b550db#diff-0aec22a6fe4107b7925c7d7fe6ab2e7f043611b830dd5a397183216e77b0dac7R257)
> 
> The code here is basically saying, "if we finished decoding before 1ms, then sleep till then." So 1ms per frame == 1000fps decoding. Now I'd like to switch this to a RageTimer too (right now it's using chrono), but I'm relying on `std::this_thread::sleep_until`, which takes a chrono time input. Perhaps I should be using a `RageThread` instead (or convert from RageTimer->chrono time), I can do a bit more research there.

Thanks for the explanation! I think it'd be great if the comment mentions that the relevant value is located in DecodeMovie.

Reply:
> Just to be sure, the FPS in this case relate to FPS of the video file to be decoded and played, not FPS rendered by the game itself, right?

Correct, or to be more precise, this is the decoding speed of the movie. Not at all tied to the framerate of ITGm or even display rate of the movie.

> Thanks for the explanation! I think it'd be great if the comment mentions that the relevant value is located in DecodeMovie.

My preference is usually to comment on the function definition itself (rather than the implementation). That way when you hover over the function calls the comment shows up as a tooltip in most IDEs.  i.e., this comment should get attached to DecodeMovie by virtue of where I placed it.